MSC3676: Transitioning away from reply fallbacks (#3676)
matrix-org/matrix-doc: proposals/3676-transitioning-away-from-reply-fallbacks.md

> # MSC3676: Transitioning away from reply fallbacks.

Reviewer: So this is essentially saying “we will go ahead with threads before removing reply fallbacks”, but why did this have to be an MSC? Is there something about reply fallbacks that threads originally was blocked on? And if so, what exactly in this proposal unblocks/decouples that tiny part? What is the essential purpose of this MSC? To me this document seems to signify intentions (partly from the Matrix foundation, SCT, and Element), which I believe is more part of the SCT process than any MSC-induced one.

Reviewer 2: > Is there something about reply fallbacks that threads originally was blocked on?

Threads want to fall back to replies, and threads want to support types that currently don't support reply fallbacks. With the current spec the reply fallbacks are required so this isn't allowed.

> what exactly in this proposal unblocks/decouples that tiny part?

Relaxing the reply fallbacks requirements to allow replying with a non-text event, which some clients have been offering for a while (e.g. I believe nheko can reply with an image/file which is outside of spec currently).

> What is the essential purpose of this MSC?

Removing reply fallbacks outright as a dependency to threads would additionally mean restructuring push rules (i.e. notification on replies) first, because essentially replies ping you by accident due to inclusion of your name in the reply fallback. This is considered a great effort, but element wants to land threads asap.

> What exactly is this proposal doing that can’t be added into the threading proposal now?

1. Keeps separate things separate, same as you would create feature branches/commits rather than just commit all of the week's work
2. By soft-deprecating fallbacks, the ecosystem is given a hint/push to implement replies properly if at all possible
3. It isn't completely decided whether fallbacks are going to be removed outright as in #2781, there are arguments to move reply fallbacks to an extra field to allow "5 lines of bash" clients to implement replies without actually implementing replies (though some might have different opinions).

Author: @HarHarLinks's points are all correct, but I think they're already captured explicitly in the problem statement for this MSC?  

In terms of Element v. Matrix.org confusion, i've made it abundantly clear in this MSC that the decision to ship threads for Matrix is **UNEQUIVOCALLY FOR THE BENEFIT OF MATRIX**.  No corners are being cut on the spec process here; and certainly none in favour of Element.  Instead, I'm just trying to avoid development of one feature getting hard-blocked unnecessarily on fixing unrelated problems elsewhere in the spec.  Meanwhile, we obviously need to fix the problems too, and we will.

   from firing notifications.
 * In the absence of fallbacks, in order to render replies simple clients will
   now have to parse `m.in_reply_to` objects and fish around for the missing
   events (or ask the server to bundle the replies, which is not yet a
   thing).

Meanwhile, [MSC3440](https://github.com/matrix-org/matrix-doc/pull/3440)
(Threads) uses replies as a fallback representation for threads (which is
very desirable to support clients while the threads transition is happening,
or to support simpler clients which support replies but not threads).
However, currently `m.in_reply_to` is only allowed on `m.room.message` events
of msgtype `m.text`, which means it cannot currently be used as a fallback
for arbitrary threaded events.

## Proposal

As a transitional step towards removing reply fallbacks entirely, instead: we
make reply fallbacks best effort.  Specifically:

 * `m.in_reply_to` is relaxed to apply to any event type
 * In practice only `m.room.message` events with msgtype `m.text` or similar
   (`m.emote`, `m.notice`) would be able to express reply fallbacks (using the
   `m.body`, `format` and `formatted_body` fields).

>  * Thread events using replies as a fallback representation for threads would
>    not include a textual reply fallback at all.

Reviewer 3: since this is really a change that needs making in MSC3440, maybe phrase it more generally:

```suggestion
 * Applications (such as threads, per MSC3440) are free not to include a textual reply fallback at all where this gives a better user experience.
```

Author: have incorporated.

This means that we can still use reply fallbacks for notification purposes
until that is properly fixed by [MSC2781](https://github.com/matrix-org/matrix-doc/pull/2781)
and [MSC3664](https://github.com/matrix-org/matrix-doc/pull/3664) - decoupling this
additional work from landing threads in

> [MSC3440](https://github.com/matrix-org/matrix-doc/pull/3440).

Reviewer 4: So the plan is to send a reply fallback in threads, if that sets the `render_in: [threads]` flag to cause a notification for those messages?

Reviewer 5: That's how I interpret this too.
If a user wants to reply to a specific event in threads, then `m.render_in: [m.thread]` will be set and the client will  add the reply fallback

Author: hang on - the intention is that threaded messages have `m.in_reply_to` but do *not* include a textual reply fallback (unless they are a reply-in-a-thread, and even then, it's best effort).  Hence the "Thread events using replies as a fallback representation for threads would not include a textual reply fallback at all." on [line 46](https://github.com/matrix-org/matrix-doc/pull/3676/files#diff-50b7eb56c09e4f77f5de28570e8ace5332469f23d0c4ad3d526fd43876b178f3R46)

Reviewer 2: I think all of you are saying the same thing: `render_in: [threads]` is the case where "they are a reply-in-a-thread, and even then, it's best effort".

Replying to a message with an attachment won't trigger a notification, but
this is no worse than the behaviour today.

> The assertion is that adding threads to Matrix is (much) higher priority and

Reviewer 3: I found this unclear.

```suggestion
We believe that adding threads to Matrix is (much) higher priority and
```

value for Matrix than cleaning up edge cases around reply fallbacks, and
given the two can be decoupled, they should be.  The importance of threads is
based on the fact that we're seeing Matrix repeatedly fail to be selected as
a collaboration technology thanks to other alternatives supporting

> Slack-style threads.

Reviewer 3: This isn't really part of the proposal - make it part of the introductory background, or a separate "justification" section?

Author: i dunno; i see the justification of the proposal as being an integral part of the proposal.

## Alternatives

We could remove fallbacks entirely and do all the subsequent work needed to
support that ([MSC2781](https://github.com/matrix-org/matrix-doc/pull/2781),
[MSC3664](https://github.com/matrix-org/matrix-doc/pull/3664) and whatever
MSC handles thread+fallback notification interaction).  However, as per
above, this feels lower priority than the importance of adding threads to

> Matrix, which would be unblocked rapidly by this MSC.

Reviewer 3: a second (not very good?) alternative might be not to use replies as a fallback for threads (somehow)?

 * In practice only `m.room.message` events with msgtype `m.text` or similar
   (`m.emote`, `m.notice`) would be able to express reply fallbacks (using the
   `m.body`, `format` and `formatted_body` fields).
 * Thread events using replies as a fallback representation for threads should
   not include a textual reply fallback at all (and so avoid threaded messages

>    triggering notifications).  The same would apply for any other usage which uses

Reviewer 2: ```diff
-   triggering notifications).  The same would apply for any other usage which uses
+   triggering notifications). The same would apply for any other usage which uses
```

Reviewer 2: There some more double whitespaces in this.

Author: life is *way* too short to worry about double-whitespace in a doc which will be viewed as HTML, which collapses it anyway. i was brought up on wordstar :P
